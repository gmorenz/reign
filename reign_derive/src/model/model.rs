use crate::{model::attr::Attr, INTERNAL_ERR};

use inflector::{cases::snakecase::to_snake_case, string::pluralize::to_plural};
use proc_macro2::{Span, TokenStream};
use proc_macro_error::{abort, abort_call_site};
use quote::quote;
use syn::{
    punctuated::Punctuated, token::Comma, Attribute, Data, DataStruct, DeriveInput, Field, Fields,
    Ident, Visibility,
};

pub fn model(input: DeriveInput) -> TokenStream {
    let DeriveInput {
        data,
        vis,
        ident,
        attrs,
        ..
    } = &input;

    match data {
        Data::Struct(DataStruct {
            fields: Fields::Named(ref fields),
            ..
        }) => gen_for_struct(Model::new(vis, ident, &fields.named, attrs)),
        _ => abort_call_site!("`#[derive(Model)]` only supports non-unit and non-tuple structs"),
    }
}

fn gen_for_struct(model: Model) -> TokenStream {
    // TODO: model: generics
    let gen_id = model.gen_id();
    let gen_filterable = model.gen_filterable();
    let gen_selectable = model.gen_selectable();
    let gen_insertable = model.gen_insertable();
    let gen_updateable = model.gen_updateable();
    let gen_deleteable = model.gen_deleteable();
    let gen_tags = model.gen_tags();

    quote! {
        #gen_id
        #gen_filterable
        #gen_selectable
        #gen_insertable
        #gen_updateable
        #gen_deleteable
        #(#gen_tags)*
    }
}

#[derive(Clone)]
pub struct ModelField {
    pub field: Field,
    pub attrs: Vec<Attr>,
    pub column_ident: Ident,
    pub no_write: bool,
    pub primary_key: bool,
    pub tags: Vec<Ident>,
}

impl ModelField {
    fn new(field: &Field, primary_keys: &[Ident]) -> Self {
        let attrs = Attr::parse_attributes(&field.attrs, false);

        let mut column_ident = field.ident.as_ref().expect(INTERNAL_ERR).clone();
        let mut no_write = false;
        let mut tags = vec![];

        for attr in &attrs {
            match attr {
                Attr::ColumnName(_, value) => column_ident = value.clone(),
                Attr::NoWrite(_) => no_write = true,
                Attr::Tag(_, value) => value.iter().for_each(|i| tags.push(i.clone())),
                _ => {}
            }
        }

        let primary_key = (primary_keys.is_empty() && column_ident == "id")
            || primary_keys.iter().find(|x| **x == column_ident).is_some();

        Self {
            field: field.to_owned(),
            attrs,
            column_ident,
            no_write,
            primary_key,
            tags,
        }
    }
}

#[derive(Clone)]
pub struct Model {
    pub vis: Visibility,
    pub ident: Ident,
    pub attrs: Vec<Attr>,
    pub fields: Vec<ModelField>,
    pub table_ident: Ident,
    pub primary_keys_size: usize,
}

impl Model {
    fn new(
        vis: &Visibility,
        ident: &Ident,
        fields: &Punctuated<Field, Comma>,
        attrs: &[Attribute],
    ) -> Self {
        let attrs = Attr::parse_attributes(attrs, true);

        let mut table_ident =
            Ident::new(&to_plural(&to_snake_case(&ident.to_string())), ident.span());
        let mut primary_keys = vec![];

        for attr in &attrs {
            match attr {
                Attr::TableName(_, value) => table_ident = value.clone(),
                Attr::PrimaryKey(_, value) => primary_keys = value.iter().cloned().collect(),
                _ => {}
            }
        }

        let fields = fields
            .into_iter()
            .map(|f| ModelField::new(f, &primary_keys))
            .collect::<Vec<_>>();

        if primary_keys.is_empty() {
            primary_keys.push(Ident::new("id", Span::call_site()));
        }

        // Check if primary key ident isn't a column
        for key in &primary_keys {
            if fields.iter().find(|x| x.column_ident == *key).is_none() {
                abort!(
                    key,
                    format!("unable to find column `{}` used for primary key", key)
                );
            }
        }

        Self {
            vis: vis.clone(),
            ident: ident.clone(),
            attrs,
            fields,
            table_ident,
            primary_keys_size: primary_keys.len(),
        }
    }

    pub fn schema(&self) -> TokenStream {
        quote! {
            crate::schema
        }
    }

    pub fn db(&self) -> TokenStream {
        quote! {
            ::reign::model::Database::get()
        }
    }

    pub fn backend(&self) -> TokenStream {
        quote! {
            ::reign::model::diesel::pg::Pg
        }
    }
}
