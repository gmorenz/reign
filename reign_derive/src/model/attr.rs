use proc_macro_error::{abort, ResultExt};
use syn::{
    parenthesized,
    parse::{Parse, ParseStream, Result},
    punctuated::Punctuated,
    token::{Comma, Eq},
    Attribute, Ident,
};

#[derive(Clone)]
pub enum Attr {
    NoInsert(Ident),
    NoUpdate(Ident),
    Tag(Ident, Punctuated<Ident, Comma>),
    ColumnName(Ident, Ident),
    TableName(Ident, Ident),
    PrimaryKey(Ident, Punctuated<Ident, Comma>),
}

impl Parse for Attr {
    fn parse(input: ParseStream) -> Result<Self> {
        let name: Ident = input.parse()?;

        match name.to_string().as_str() {
            "no_insert" => Ok(Attr::NoInsert(name)),
            "no_update" => Ok(Attr::NoUpdate(name)),
            "tag" => Ok(Attr::Tag(name, parenthesized_list(input)?)),
            "column_name" => Ok(Attr::ColumnName(name, eq(input)?)),
            "table_name" => Ok(Attr::TableName(name, eq(input)?)),
            "primary_key" => Ok(Attr::PrimaryKey(name, parenthesized_list(input)?)),
            _ => abort!(name, "unexpected attribute: {}", name),
        }
    }
}

impl Attr {
    pub fn parse_attributes(attrs: &[Attribute], for_struct: bool) -> Vec<Self> {
        let attrs = attrs
            .iter()
            .filter(|attr| attr.path.is_ident("model"))
            .flat_map(|attr| {
                attr.parse_args_with(Punctuated::<Attr, Comma>::parse_terminated)
                    .unwrap_or_abort()
            })
            .collect::<Vec<_>>();

        for attr in &attrs {
            match attr {
                Attr::NoInsert(ident) if for_struct => {
                    abort!(ident, "`no_insert` is not allowed on struct")
                }
                Attr::NoUpdate(ident) if for_struct => {
                    abort!(ident, "`no_update` is not allowed on struct")
                }
                Attr::Tag(ident, _) if for_struct => {
                    abort!(ident, "`tag` is not allowed on struct")
                }
                Attr::ColumnName(ident, _) if for_struct => {
                    abort!(ident, "`column_name` is not allowed on struct")
                }
                Attr::TableName(ident, _) if !for_struct => {
                    abort!(ident, "`table_name` is not allowed on field")
                }
                Attr::PrimaryKey(ident, _) if !for_struct => {
                    abort!(ident, "`primary_key` is not allowed on field")
                }
                _ => {}
            }
        }

        attrs
    }
}

fn eq<T: Parse>(input: ParseStream) -> Result<T> {
    input.parse::<Eq>()?;
    input.parse()
}

fn parenthesized_list<T: Parse>(input: ParseStream) -> Result<Punctuated<T, Comma>> {
    let content;
    parenthesized!(content in input);

    content.parse_terminated(T::parse)
}
