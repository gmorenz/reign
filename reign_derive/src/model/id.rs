use crate::{
    model::model::{Model, ModelField},
    INTERNAL_ERR,
};

use proc_macro2::TokenStream;
use quote::quote;
use syn::Ident;

impl Model {
    pub fn gen_id(&self) -> TokenStream {
        let gen_identifiable = self.gen_identifiable(&self.ident, &self.fields);

        quote! {
            #gen_identifiable
        }
    }

    pub fn gen_tag_id(&self, ident: &Ident, fields: &[ModelField]) -> TokenStream {
        let gen_tag_identifiable = self.gen_identifiable(ident, fields);

        quote! {
            #gen_tag_identifiable
        }
    }

    // Generates `id` method
    fn gen_identifiable(&self, ident: &Ident, fields: &[ModelField]) -> TokenStream {
        let table_ident = &self.table_ident;
        let schema = self.schema();

        let keys = fields.iter().filter(|x| x.primary_key).collect::<Vec<_>>();

        if keys.len() != self.primary_keys_size {
            return quote! {};
        }

        let (field_ty, field_ident) = keys
            .iter()
            .map(|f| (&f.field.ty, f.field.ident.as_ref().expect(INTERNAL_ERR)))
            .unzip::<_, _, Vec<_>, Vec<_>>();

        quote! {
            impl ::reign::model::diesel::associations::HasTable for #ident
            {
                type Table = #schema::#table_ident::table;

                fn table() -> Self::Table {
                    #schema::#table_ident::table
                }
            }

            impl<'ident> ::reign::model::diesel::associations::Identifiable for &'ident #ident
            {
                type Id = (#(&'ident #field_ty),*);

                fn id(self) -> Self::Id {
                    (#(&self.#field_ident),*)
                }
            }
        }
    }
}
