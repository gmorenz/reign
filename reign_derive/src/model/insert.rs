use crate::{
    model::model::{Model, ModelField},
    INTERNAL_ERR,
};

use proc_macro2::TokenStream;
use quote::{format_ident, quote};
use syn::{Field, Ident};

impl Model {
    pub fn gen_insert(&self) -> TokenStream {
        let gen_insert_struct = self.gen_insert_struct();
        let gen_insert_setters = self.gen_insert_setters();
        let gen_insert_methods = self.gen_insert_methods(&self.ident);
        let gen_insert_action = self.gen_insert_action(&self.ident, &self.fields);

        quote! {
            #gen_insert_struct
            #gen_insert_setters
            #gen_insert_methods
            #gen_insert_action
        }
    }

    pub fn gen_tag_insert(&self, ident: &Ident, fields: &[ModelField]) -> TokenStream {
        let gen_tag_insert_methods = self.gen_insert_methods(ident);
        let gen_tag_insert_action = self.gen_insert_action(ident, fields);

        quote! {
            #gen_tag_insert_methods
            #gen_tag_insert_action
        }
    }

    fn insert_ident(&self) -> Ident {
        format_ident!("Insert{}", self.ident)
    }

    // Generates struct & constructor for `INSERT` statements
    fn gen_insert_struct(&self) -> TokenStream {
        let insert_ident = self.insert_ident();
        let vis = &self.vis;

        let (for_struct, for_new) = self
            .fields
            .iter()
            .filter(|x| !x.no_insert)
            .map(|x| {
                let Field { vis, ty, ident, .. } = &x.field;
                let ident = ident.as_ref().expect(INTERNAL_ERR);

                (
                    quote! {
                        #vis #ident: Option<#ty>
                    },
                    quote! {
                        #ident: None
                    },
                )
            })
            .unzip::<_, _, Vec<_>, Vec<_>>();

        quote! {
            #vis struct #insert_ident<M> {
                _phantom: std::marker::PhantomData<(M)>,
                #(#for_struct,)*
            }

            impl<M> #insert_ident<M> {
                fn new() -> Self {
                    Self {
                        _phantom: std::marker::PhantomData,
                        #(#for_new,)*
                    }
                }
            }
        }
    }

    // Generates individual column setters for `INSERT`
    fn gen_insert_setters(&self) -> TokenStream {
        let insert_ident = self.insert_ident();
        // let schema = self.schema();
        // let backend = self.backend();

        // let (field_vis, field_ident) = self
        //     .fields
        //     .iter()
        //     .filter(|x| !x.no_insert)
        //     .map(|x| (&x.field.vis, x.field.ident.as_ref().expect(INTERNAL_ERR)))
        //     .unzip::<_, _, Vec<_>, Vec<_>>();

        quote! {
            impl<M> #insert_ident<M> {
                // #(#field_vis fn #field_ident<E, X>(mut self, #field_ident: E) -> Self
                // {
                //     self.#field_ident = Some(#field_ident);
                //     self
                // })*
            }
        }
    }

    // Generates starting methods for `INSERT`
    fn gen_insert_methods(&self, ident: &Ident) -> TokenStream {
        let insert_ident = self.insert_ident();
        let vis = &self.vis;

        quote! {
            #[allow(dead_code, unreachable_code)]
            impl #ident {
                #vis fn new() -> #insert_ident<#ident> {
                    #insert_ident::new()
                }
            }
        }
    }

    // Generates actual action for `INSERT`
    fn gen_insert_action(&self, ident: &Ident, fields: &[ModelField]) -> TokenStream {
        let insert_ident = self.insert_ident();
        let table_ident = &self.table_ident;
        let schema = self.schema();
        let db = self.db();
        let vis = &self.vis;

        let field_ident = fields
            .iter()
            .map(|x| x.field.ident.as_ref().expect(INTERNAL_ERR))
            .collect::<Vec<_>>();

        quote! {
            impl #insert_ident<#ident> {
                #vis async fn save(self) -> Result<#ident, ::reign::model::tokio_diesel::AsyncError> {
                    use ::reign::model::tokio_diesel::AsyncRunQueryDsl;
                    use ::reign::model::diesel::ExpressionMethods;

                    ::reign::model::diesel::insert_into(#schema::#table_ident::table)
                        .values((
                            #schema::#table_ident::id.eq(1),
                        ))
                        .returning((
                            #(#schema::#table_ident::#field_ident,)*
                        ))
                        .get_result_async::<#ident>(#db)
                        .await
                }
            }
        }
    }
}
