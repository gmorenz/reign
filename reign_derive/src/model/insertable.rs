use crate::{
    model::model::{Model, ModelField},
    INTERNAL_ERR,
};

use proc_macro2::TokenStream;
use quote::{format_ident, quote};
use syn::{Field, Ident};

// TODO: model: Allow batch inserting using `Vec`
impl Model {
    pub fn gen_insertable(&self) -> TokenStream {
        let gen_insertable_struct = self.gen_insertable_struct();
        let gen_insertable_trait = self.gen_insertable_trait();
        let gen_insertable_setters = self.gen_insertable_setters();
        let gen_insertable_methods = self.gen_insertable_methods(&self.ident);
        let gen_insertable_actions = self.gen_insertable_actions(&self.ident, &self.fields);

        quote! {
            #gen_insertable_struct
            #gen_insertable_trait
            #gen_insertable_setters
            #gen_insertable_methods
            #gen_insertable_actions
        }
    }

    pub fn gen_tag_insertable(&self, ident: &Ident, fields: &[ModelField]) -> TokenStream {
        let gen_tag_insertable_methods = self.gen_insertable_methods(ident);
        let gen_tag_insertable_actions = self.gen_insertable_actions(ident, fields);

        quote! {
            #gen_tag_insertable_methods
            #gen_tag_insertable_actions
        }
    }

    fn insertable_ident(&self) -> Ident {
        format_ident!("Insertable{}", self.ident)
    }

    // Generates struct & constructor for `INSERT` statements
    fn gen_insertable_struct(&self) -> TokenStream {
        let insertable_ident = self.insertable_ident();
        let vis = &self.vis;

        let (for_struct, for_new) = self
            .fields
            .iter()
            .filter(|x| !x.no_write)
            .map(|f| {
                let Field { ty, ident, .. } = &f.field;
                let ident = ident.as_ref().expect(INTERNAL_ERR);

                (
                    quote! {
                        #ident: Option<#ty>
                    },
                    quote! {
                        #ident: None
                    },
                )
            })
            .unzip::<_, _, Vec<_>, Vec<_>>();

        quote! {
            #vis struct #insertable_ident<M> {
                _phantom: std::marker::PhantomData<(M)>,
                #(#for_struct,)*
            }

            impl<M> #insertable_ident<M> {
                fn new() -> Self {
                    Self {
                        _phantom: std::marker::PhantomData,
                        #(#for_new,)*
                    }
                }
            }
        }
    }

    fn gen_insertable_trait(&self) -> TokenStream {
        let insertable_ident = self.insertable_ident();
        let table_ident = &self.table_ident;
        let schema = self.schema();

        let (val_ty, val) = self
            .fields
            .iter()
            .filter(|x| !x.no_write)
            .map(|f| {
                let Field { ident, ty, .. } = &f.field;
                let ident = ident.as_ref().expect(INTERNAL_ERR);
                let column_ident = &f.column_ident;

                (
                    quote! {
                        Option<::reign::model::diesel::dsl::Eq<#schema::#table_ident::#column_ident, #ty>>
                    },
                    quote! {
                        self.#ident.map(|x| #schema::#table_ident::#column_ident.eq(x))
                    },
                )
            })
            .unzip::<_, _, Vec<_>, Vec<_>>();

        quote! {
            impl<M> ::reign::model::diesel::Insertable<#schema::#table_ident::table> for #insertable_ident<M>
            {
                type Values = <(#(#val_ty,)*) as ::reign::model::diesel::Insertable<#schema::#table_ident::table>>::Values;

                fn values(self) -> Self::Values {
                    use ::reign::model::diesel::ExpressionMethods;

                    (#(#val,)*).values()
                }
            }
        }
    }

    // Generates individual column setters for `INSERT`
    // TODO:(model) Allow `AsExpression<SqlTypeOf>` so that we can take any value
    fn gen_insertable_setters(&self) -> TokenStream {
        let insertable_ident = self.insertable_ident();

        let setters = self
            .fields
            .iter()
            .filter(|x| !x.no_write)
            .map(|f| {
                let Field { vis, ident, ty, .. } = &f.field;
                let ident = ident.as_ref().expect(INTERNAL_ERR);

                quote! {
                    #vis fn #ident(mut self, #ident: #ty) -> Self {
                        self.#ident = Some(#ident);
                        self
                    }
                }
            })
            .collect::<Vec<_>>();

        quote! {
            impl<M> #insertable_ident<M> {
                #(#setters)*
            }
        }
    }

    // Generates starting methods for `INSERT`
    fn gen_insertable_methods(&self, ident: &Ident) -> TokenStream {
        let insertable_ident = self.insertable_ident();
        let vis = &self.vis;

        quote! {
            #[allow(dead_code, unreachable_code)]
            impl #ident {
                #vis fn new() -> #insertable_ident<#ident> {
                    #insertable_ident::new()
                }
            }
        }
    }

    // Generates actual action for `INSERT`
    fn gen_insertable_actions(&self, ident: &Ident, fields: &[ModelField]) -> TokenStream {
        let insertable_ident = self.insertable_ident();
        let table_ident = &self.table_ident;
        let schema = self.schema();
        let db = self.db();
        let vis = &self.vis;

        let column_ident = fields.iter().map(|x| &x.column_ident).collect::<Vec<_>>();

        quote! {
            impl #insertable_ident<#ident> {
                #vis async fn save(self) -> Result<#ident, ::reign::model::Error> {
                    use ::reign::model::tokio_diesel::AsyncRunQueryDsl;
                    use ::reign::model::diesel::ExpressionMethods;

                    Ok(::reign::model::diesel::insert_into(#schema::#table_ident::table)
                        .values(self)
                        .returning((
                            #(#schema::#table_ident::#column_ident,)*
                        ))
                        .get_result_async::<#ident>(#db)
                        .await?)
                }
            }
        }
    }
}
