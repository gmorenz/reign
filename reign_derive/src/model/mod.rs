#[allow(clippy::module_inception)]
pub mod model;

mod attr;
mod id;
mod insert;
mod query;
mod tag;
mod update;
