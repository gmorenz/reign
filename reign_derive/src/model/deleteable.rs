use crate::{
    model::model::{Model, ModelField},
    INTERNAL_ERR,
};

use proc_macro2::TokenStream;
use quote::{format_ident, quote};
use syn::Ident;

impl Model {
    pub fn gen_deleteable(&self) -> TokenStream {
        let gen_deleteable_struct = self.gen_deleteable_struct();
        let gen_deleteable_filters = self.gen_deleteable_filters();
        let gen_deleteable_methods = self.gen_deleteable_methods(&self.ident, &self.fields);
        let gen_deleteable_actions = self.gen_deleteable_actions(&self.ident, &self.fields);

        quote! {
            #gen_deleteable_struct
            #gen_deleteable_filters
            #gen_deleteable_methods
            #gen_deleteable_actions
        }
    }

    pub fn gen_tag_deleteable(&self, ident: &Ident, fields: &[ModelField]) -> TokenStream {
        let gen_deleteable_methods = self.gen_deleteable_methods(ident, fields);
        let gen_deleteable_actions = self.gen_deleteable_actions(ident, fields);

        quote! {
            #gen_deleteable_methods
            #gen_deleteable_actions
        }
    }

    fn deleteable_ident(&self) -> Ident {
        format_ident!("Deleteable{}", self.ident)
    }

    // Generates struct & constructor for `DELETE` statements
    fn gen_deleteable_struct(&self) -> TokenStream {
        let deleteable_ident = self.deleteable_ident();
        let vis = &self.vis;
        let table_ident = &self.table_ident;
        let schema = self.schema();
        let backend = self.backend();

        quote! {
            #vis struct #deleteable_ident<M> {
                _phantom: std::marker::PhantomData<(M)>,
                statement: ::reign::model::diesel::query_builder::BoxedDeleteStatement<
                    'static,
                    #backend,
                    #schema::#table_ident::table
                >,
            }

            impl<M> #deleteable_ident<M> {
                fn new() -> Self {
                    Self {
                        _phantom: std::marker::PhantomData,
                        statement: ::reign::model::diesel::delete(#schema::#table_ident::table).into_boxed(),
                    }
                }
            }
        }
    }

    // Generates individual column filters for `DELETE`
    fn gen_deleteable_filters(&self) -> TokenStream {
        let deleteable_ident = self.deleteable_ident();
        let table_ident = &self.table_ident;
        let schema = self.schema();
        let backend = self.backend();

        let (field_vis, column_ident) = self
            .fields
            .iter()
            .map(|x| (&x.field.vis, &x.column_ident))
            .unzip::<_, _, Vec<_>, Vec<_>>();

        quote! {
            #[allow(dead_code, unreachable_code)]
            impl<M> #deleteable_ident<M> {
                #(#field_vis fn #column_ident<E, X>(mut self, #column_ident: E) -> Self
                where
                    E: ::reign::model::diesel::expression::AsExpression<
                        ::reign::model::diesel::dsl::SqlTypeOf<#schema::#table_ident::#column_ident>,
                        Expression = X,
                    >,
                    X: ::reign::model::diesel::expression::BoxableExpression<
                            #schema::#table_ident::table,
                            #backend,
                            SqlType = ::reign::model::diesel::dsl::SqlTypeOf<#schema::#table_ident::#column_ident>
                        >
                        + ::reign::model::diesel::expression::ValidGrouping<
                            (),
                            IsAggregate = ::reign::model::diesel::expression::is_aggregate::Never
                        >
                        + Send
                        + 'static,
                {
                    use ::reign::model::diesel::{ExpressionMethods, QueryDsl};

                    self.statement = self.statement.filter(#schema::#table_ident::#column_ident.eq(#column_ident));
                    self
                })*
            }
        }
    }

    // Generates starting methods for `UPDATE`
    fn gen_deleteable_methods(&self, ident: &Ident, fields: &[ModelField]) -> TokenStream {
        let deleteable_ident = self.deleteable_ident();
        let vis = &self.vis;

        let keys = fields.iter().filter(|x| x.primary_key).collect::<Vec<_>>();

        if keys.len() != self.primary_keys_size {
            return quote! {};
        }

        let (column_ident, field_ident) = keys
            .iter()
            .map(|f| (&f.column_ident, f.field.ident.as_ref().expect(INTERNAL_ERR)))
            .unzip::<_, _, Vec<_>, Vec<_>>();

        quote! {
            #[allow(dead_code, unreachable_code)]
            impl #ident {
                #vis fn delete() -> #deleteable_ident<Vec<#ident>> {
                    #deleteable_ident::new()
                }

                #vis async fn drop(&self) -> Result<#ident, ::reign::model::Error> {
                    #deleteable_ident::<#ident>::new()
                        #(.#column_ident(self.#field_ident.clone()))*
                        .drop()
                        .await
                }
            }
        }
    }

    // Generates actual action for `DELETE`
    fn gen_deleteable_actions(&self, ident: &Ident, fields: &[ModelField]) -> TokenStream {
        let deleteable_ident = self.deleteable_ident();
        let table_ident = &self.table_ident;
        let schema = self.schema();
        let db = self.db();
        let vis = &self.vis;

        let column_ident = fields.iter().map(|x| &x.column_ident).collect::<Vec<_>>();

        quote! {
            #[allow(dead_code, unreachable_code)]
            impl #deleteable_ident<Vec<#ident>> {
                #vis async fn drop(self) -> Result<Vec<#ident>, ::reign::model::Error> {
                    use ::reign::model::tokio_diesel::AsyncRunQueryDsl;
                    use ::reign::model::diesel::QueryDsl;

                    Ok(self.statement
                        .returning((
                            #(#schema::#table_ident::#column_ident,)*
                        ))
                        .get_results_async::<#ident>(#db)
                        .await?)
                }
            }

            #[allow(dead_code, unreachable_code)]
            impl #deleteable_ident<#ident> {
                #vis async fn drop(self) -> Result<#ident, ::reign::model::Error> {
                    use ::reign::model::tokio_diesel::AsyncRunQueryDsl;
                    use ::reign::model::diesel::QueryDsl;

                    Ok(self.statement
                        .returning((
                            #(#schema::#table_ident::#column_ident,)*
                        ))
                        .get_result_async::<#ident>(#db)
                        .await?)
                }
            }
        }
    }
}
