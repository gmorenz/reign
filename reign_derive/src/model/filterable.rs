use crate::model::model::{Model, ModelField};

use proc_macro2::TokenStream;
use quote::{format_ident, quote};
use syn::Ident;

impl Model {
    pub fn gen_filterable(&self) -> TokenStream {
        let gen_filterable_struct = self.gen_filterable_struct();
        let gen_filterable_filters = self.gen_filterable_filters();
        let gen_filterable_methods = self.gen_filterable_methods(&self.ident);

        quote! {
            #gen_filterable_struct
            #gen_filterable_filters
            #gen_filterable_methods
        }
    }

    pub fn gen_tag_filterable(&self, ident: &Ident, _fields: &[ModelField]) -> TokenStream {
        let gen_filterable_methods = self.gen_filterable_methods(ident);

        quote! {
            #gen_filterable_methods
        }
    }

    pub fn filterable_ident(&self) -> Ident {
        format_ident!("Filterable{}", self.ident)
    }

    fn gen_filterable_struct(&self) -> TokenStream {
        let filterable_ident = self.filterable_ident();
        let vis = &self.vis;
        let table_ident = &self.table_ident;
        let schema = self.schema();
        let backend = self.backend();

        // TODO: external: Use dummy mod once https://github.com/rust-analyzer/rust-analyzer/issues/1559
        quote! {
            #vis struct #filterable_ident<M> {
                _phantom: std::marker::PhantomData<(M)>,
                statement: Box<
                    dyn ::reign::model::diesel::expression::BoxableExpression<
                        #schema::#table_ident::table,
                        #backend,
                        SqlType = ::reign::model::diesel::sql_types::Bool,
                    >,
                >,
            }

            impl<M> #filterable_ident<M> {
                fn new() -> Self {
                    use ::reign::model::diesel::{
                        sql_types::{Bool, Nullable},
                        PgExpressionMethods, IntoSql,
                    };

                    let none: Option<bool> = None;

                    Self {
                        _phantom: std::marker::PhantomData,
                        statement: Box::new(none.into_sql::<Nullable<Bool>>().is_not_distinct_from(none)),
                    }
                }
            }
        }
    }

    // TODO: model: More filters when both `Bool` and `Nullable<Bool>` can be stored at once
    // Generates individual column filters
    fn gen_filterable_filters(&self) -> TokenStream {
        let filterable_ident = self.filterable_ident();
        let table_ident = &self.table_ident;
        let schema = self.schema();
        let backend = self.backend();

        let (field_vis, column_ident) = self
            .fields
            .iter()
            .map(|x| (&x.field.vis, &x.column_ident))
            .unzip::<_, _, Vec<_>, Vec<_>>();

        quote! {
            #[allow(dead_code, unreachable_code)]
            impl<M> #filterable_ident<M> {
                #(#field_vis fn #column_ident<E, X>(mut self, #column_ident: E) -> Self
                where
                    E: ::reign::model::diesel::expression::AsExpression<
                        ::reign::model::diesel::dsl::SqlTypeOf<#schema::#table_ident::#column_ident>,
                        Expression = X,
                    >,
                    X: ::reign::model::diesel::expression::BoxableExpression<
                            #schema::#table_ident::table,
                            #backend,
                            SqlType = ::reign::model::diesel::dsl::SqlTypeOf<#schema::#table_ident::#column_ident>
                        >
                        + ::reign::model::diesel::expression::ValidGrouping<
                            (),
                            IsAggregate = ::reign::model::diesel::expression::is_aggregate::Never
                        >
                        + Send
                        + 'static,
                {
                    use ::reign::model::diesel::{BoolExpressionMethods, PgExpressionMethods, QueryDsl};

                    self.statement = Box::new(self.statement.and(#schema::#table_ident::#column_ident.is_not_distinct_from::<E>(#column_ident)));
                    self
                })*
            }
        }
    }

    fn gen_filterable_methods(&self, ident: &Ident) -> TokenStream {
        let filterable_ident = self.filterable_ident();
        let vis = &self.vis;

        quote! {
            #[allow(dead_code, unreachable_code)]
            impl #ident {
                #vis fn filter() -> #filterable_ident<#ident> {
                    #filterable_ident::new()
                }
            }
        }
    }
}
