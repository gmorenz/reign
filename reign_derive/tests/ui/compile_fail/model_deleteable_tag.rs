use reign::{model::diesel, prelude::*};

mod schema {
    use reign::model::diesel;

    diesel::table! {
        users (id) {
            id -> Int4,
            name -> Varchar,
        }
    }
}

#[derive(Debug, Model)]
pub struct User {
    #[model(no_write)]
    id: i32,
    #[model(tag(name))]
    name: String,
}

#[tokio::main]
async fn main() {
    let user = UserName::one().await.unwrap();

    assert!(user.is_some());

    let user = user.unwrap();

    user.drop().await.unwrap();
}
