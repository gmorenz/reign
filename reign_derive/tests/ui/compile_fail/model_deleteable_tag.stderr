error[E0599]: no method named `drop` found for struct `UserName` in the current scope
   --> $DIR/model_deleteable_tag.rs:30:10
    |
14  | #[derive(Debug, Model)]
    |                 ----- method `drop` not found for this
...
30  |     user.drop().await.unwrap();
    |          ^^^^ method not found in `UserName`
    |
   ::: /Users/pksunkara/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/src/rust/library/core/src/ops/drop.rs:164:8
    |
164 |     fn drop(&mut self);
    |        ----
    |        |
    |        the method is available for `Box<UserName>` here
    |        the method is available for `Arc<UserName>` here
    |        the method is available for `Rc<UserName>` here
    |
    = help: items from traits can only be used if the trait is implemented and in scope
    = note: the following traits define an item `drop`, perhaps you need to implement one of them:
            candidate #1: `Drop`
            candidate #2: `UnsafeFutureObj`
            candidate #3: `pin_project::__private::PinnedDrop`
            candidate #4: `pin_project::__private::PinnedDrop`
